### DISCUSSION TITLE
CosmosDiagnostics: Expose regions

### DISCUSSION THREAD

**Discussion by j82w** [Azure/azure-cosmos-dotnet-v3#2141] [Ideas]
User's need a way to log and programmatically figure out if a request failed over to a different region. 

Suggested solution:
Add the following API to the [CosmosDiagnostics](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs).
1. Try is allow for scenario where no region succeeded.
2. IReadOnlyList is for query or other operations that could occur across multiple regions
3. This is done on the request level since client level is ambiguous. It's possible there are race conditions or only a single request failed over instead of the entire SDK instance.

```c#
public abstract class CosmosDiagnostics
{
        public abstract bool TryGetSuccessRegion(out IReadOnlyList<string> regions);

```

Updated solution based on discussion:

```c#
public abstract class CosmosDiagnostics
{
        public abstract (string region, Uri endpoint) GetContatedRegions();

```

**Comment by ealsur** [comment]
Why only the successes? Would it be good to return all contacted regions? If there was a failure (for example, a 404), which is not catastrophic, it would be good to also see which region returned the 404, right?

**Reply by j82w** [reply]
This allows the user to see what region the request actually completed on which is what really matters. With this user's can add flags and validation that the request is not going to the default region. 

The problem I have with exposing a list is ambiguous, and doesn't help users. They get to see it failed in 3 regions, but they don't have any way of seeing why it failed or how long each request took. At that point it would require the entire diagnostics object model to be exposed to actually be useful.

**Reply by kirankumarkolli** [reply]
List as contract it-self might be un-intuitive 

> might not understand the multi NW interaction and possibility)
> Order significance 

If the ask is to know which regions are contracted, then all regions do makes sense.

**Reply by j82w** [reply]
What the use case for needing to know just a list of contact regions?

**Reply by ealsur** [reply]
If I open the Diagnostics and see that it contacted 2 regions, and there were 2 requests, and the second region succeeded (for example, 404/1002 Read Session Not Available) and I only get the successful region, then (if I didn't look at the diagnostics) could assume that there was only 1 region involved, but that was not the case, right?

There are customers who track cross-region requests out of the Diagnostics, so they would see a discrepancy between what the API returned and the full detail.

Also, what happens if the request failed completely, would the API return nothing since it failed?

**Reply by j82w** [reply]
That's a good point. What about this?

`public abstract (string region, Uri endpoint) GetContatedRegions();`

**Reply by kirankumarkolli** [reply]
+1 for API which lists all the regions contacted.

**Comment by Castaly** [comment]
Good idea.

Would be good to return standard Azure region name (West US) instead of (or in addition to) physical node name like cdb-ms-prod-westus1-fdxx.documents.azure.com.
